ksqldb-engine/src/main/java/io/confluent/ksql/function/UdafFactoryInvoker.java

```diff
@@ -78,24 +81,42 @@
     this.description = Objects.requireNonNull(description);
   }
 
+  @SuppressFBWarnings({"EXS_EXCEPTION_SOFTENING_NO_CONSTRAINTS", "REC_CATCH_EXCEPTION"})
   @SuppressWarnings("unchecked")
-  KsqlAggregateFunction createFunction(final AggregateFunctionInitArguments initArgs) {
+  KsqlAggregateFunction createFunction(final AggregateFunctionInitArguments initArgs,
+      final List<SqlArgument> argTypeList) {
     final Object[] factoryArgs = initArgs.args().toArray();
     try {
       final Udaf udaf = (Udaf)method.invoke(null, factoryArgs);
 
+      udaf.initializeTypeArguments(argTypeList);
       if (udaf instanceof Configurable) {
         ((Configurable) udaf).configure(initArgs.config());
       }
 
+      final SqlType aggregateSqlType;
+      final SqlType returnSqlType;
+
+      if (udaf.getAggregateSqlType().isPresent()) {
+        aggregateSqlType = (SqlType) udaf.getAggregateSqlType().get();
+      } else {
+        aggregateSqlType = SchemaConverters.functionToSqlConverter().toSqlType(aggregateArgType);
+      }
```

Comment:
(optional) A trick using optional values:

```suggestion
aggregateSqlType = udaf.getAggregateSqlType()
   .orElse(SchemaConverters.functionToSqlConverter().toSqlType(aggregateArgType));    
```